server.c

```diff
@@ -3854,10 +3986,136 @@ handle_tcp_reading(int fd, short event, void* arg)
 
 	assert((event & EV_READ));
 
-	if (data->bytes_transmitted == 0) {
+	if (data->bytes_transmitted == 0 &&
+		data->pp2_header_state == pp2_header_none) {
 		query_reset(data->query, TCP_MAX_MESSAGE_LEN, 1);
 	}
 
+	if(data->pp2_enabled && data->pp2_header_state != pp2_header_done) {
+		struct pp2_header* header = NULL;
+		size_t want_read_size = 0;
+		size_t current_read_size = 0;
+		if(data->pp2_header_state == pp2_header_none) {
+			want_read_size = PP2_HEADER_SIZE;
+			if(buffer_remaining(data->query->packet) <
+				want_read_size) {
+				VERBOSITY(6, (LOG_ERR, "proxy-protocol: not enough buffer size to read PROXYv2 header"));
+				cleanup_tcp_handler(data);
+				return;
+			}
+			VERBOSITY(6, (LOG_INFO, "proxy-protocol: reading fixed part of PROXYv2 header (len %lu)", (unsigned long)want_read_size));
+			current_read_size = want_read_size;
+			if(data->bytes_transmitted < current_read_size) {
+				received = read(fd,
+					(void*)buffer_at(data->query->packet,
+						data->bytes_transmitted),
+						current_read_size - data->bytes_transmitted);
+				if (received == -1) {
+					if (errno == EAGAIN || errno == EINTR) {
+						/*
+						 * Read would block, wait until more
+						 * data is available.
+						 */
+						return;
+					} else {
+						char buf[48];
+						addr2str(&data->query->remote_addr, buf, sizeof(buf));
+#ifdef ECONNRESET
+						if (verbosity >= 2 || errno != ECONNRESET)
+#endif /* ECONNRESET */
+						log_msg(LOG_ERR, "failed reading from %s tcp: %s", buf, strerror(errno));
+						cleanup_tcp_handler(data);
+						return;
+					}
+				} else if (received == 0) {
+					/* EOF */
+					cleanup_tcp_handler(data);
+					return;
+				}
+
+				data->bytes_transmitted += received;
+				buffer_skip(data->query->packet, received);
+				if(data->bytes_transmitted != current_read_size)
+					return;
+				data->pp2_header_state = pp2_header_init;
+			}
+		}
+		if(data->pp2_header_state == pp2_header_init) {
+			int err;
+			err = pp2_read_header(buffer_begin(data->query->packet),
+				buffer_limit(data->query->packet));
+			if(err) {
+				VERBOSITY(6, (LOG_ERR, "proxy-protocol: could not parse PROXYv2 header: %s", pp_lookup_error(err)));
+				cleanup_tcp_handler(data);
+				return;
+			}
+			header = (struct pp2_header*)buffer_begin(data->query->packet);
+			want_read_size = ntohs(header->len);
+			if(buffer_limit(data->query->packet) <
+				PP2_HEADER_SIZE + want_read_size) {
+				VERBOSITY(6, (LOG_ERR, "proxy-protocol: not enough buffer size to read PROXYv2 header"));
+				cleanup_tcp_handler(data);
+				return;
+			}
+			VERBOSITY(6, (LOG_INFO, "proxy-protocol: reading variable part of PROXYv2 header (len %lu)", (unsigned long)want_read_size));
+			current_read_size = PP2_HEADER_SIZE + want_read_size;
+			if(want_read_size == 0) {
+				/* nothing more to read; header is complete */
+				data->pp2_header_state = pp2_header_done;
+			} else if(data->bytes_transmitted < current_read_size) {
+				received = read(fd,
```

Comment:
Seems we could benefit from implementing a function to read+handle errors instead of repeating the same? Can be postponed of course, but seems convenient

Reply:
It is in commit https://github.com/NLnetLabs/nsd/pull/281/commits/db581ac35ed36ee59694e967e313f949da676862 . That creates functions for TCP and TLS to read more data and handle an error.